cluster-autoscaler/simulator/basic_cluster_snapshot.go

```diff
@@ -0,0 +1,263 @@
+type BasicClusterSnapshot struct {
+	baseData   *internalBasicSnapshotData
+	forkedData *internalBasicSnapshotData
+}
+
+type internalBasicSnapshotDataNodeLister internalBasicSnapshotData
+type internalBasicSnapshotDataPodLister internalBasicSnapshotData
+
+type internalBasicSnapshotData struct {
+	nodeInfoMap map[string]*schedulernodeinfo.NodeInfo
+}
+
+func (data *internalBasicSnapshotDataNodeLister) List() ([]*schedulernodeinfo.NodeInfo, error) {
+	nodeInfoList := make([]*schedulernodeinfo.NodeInfo, 0, len(data.nodeInfoMap))
+	for _, v := range data.nodeInfoMap {
+		nodeInfoList = append(nodeInfoList, v)
+	}
+	return nodeInfoList, nil
+}
+
+func (data *internalBasicSnapshotDataNodeLister) HavePodsWithAffinityList() ([]*schedulernodeinfo.NodeInfo, error) {
+	havePodsWithAffinityList := make([]*schedulernodeinfo.NodeInfo, 0, len(data.nodeInfoMap))
+	for _, v := range data.nodeInfoMap {
+		if len(v.PodsWithAffinity()) > 0 {
+			havePodsWithAffinityList = append(havePodsWithAffinityList, v)
+		}
+	}
+	return havePodsWithAffinityList, nil
+}
+
+func (data *internalBasicSnapshotDataNodeLister) Get(nodeName string) (*schedulernodeinfo.NodeInfo, error) {
+	if v, ok := data.nodeInfoMap[nodeName]; ok {
+		return v, nil
+	}
+	return nil, fmt.Errorf("node %s not in snapshot", nodeName)
+}
+
+func (data *internalBasicSnapshotDataPodLister) List(selector labels.Selector) ([]*apiv1.Pod, error) {
+	alwaysTrue := func(p *apiv1.Pod) bool { return true }
+	return data.FilteredList(alwaysTrue, selector)
+}
+
+func (data *internalBasicSnapshotDataPodLister) FilteredList(podFilter schedulerlisters.PodFilter, selector labels.Selector) ([]*apiv1.Pod, error) {
+	pods := make([]*apiv1.Pod, 0)
+	for _, n := range data.nodeInfoMap {
+		for _, pod := range n.Pods() {
+			if podFilter(pod) && selector.Matches(labels.Set(pod.Labels)) {
+				pods = append(pods, pod)
+			}
+		}
+	}
+	return pods, nil
+}
+
+func (data *internalBasicSnapshotData) Pods() schedulerlisters.PodLister {
+	return (*internalBasicSnapshotDataPodLister)(data)
+}
+
+func (data *internalBasicSnapshotData) NodeInfos() schedulerlisters.NodeInfoLister {
+	return (*internalBasicSnapshotDataNodeLister)(data)
+}
+
+// NewEmptySnapshot initializes a Snapshot struct and returns it.
+func newInternalBasicSnapshotData() *internalBasicSnapshotData {
+	return &internalBasicSnapshotData{
+		nodeInfoMap: make(map[string]*schedulernodeinfo.NodeInfo),
+	}
+}
+
+func (data *internalBasicSnapshotData) clone() *internalBasicSnapshotData {
+	clonedNodeInforMap := make(map[string]*schedulernodeinfo.NodeInfo)
+	for k, v := range data.nodeInfoMap {
+		clonedNodeInforMap[k] = v.Clone()
+	}
+	return &internalBasicSnapshotData{
+		nodeInfoMap: clonedNodeInforMap,
+	}
+}
+
+func (data *internalBasicSnapshotData) addNode(node *apiv1.Node) error {
+	if _, found := data.nodeInfoMap[node.Name]; found {
+		return fmt.Errorf("node %s already in snapshot", node.Name)
```

Comment:
I imagine that would be a useful error to check against (so you can retry with a different name if it's an in-memory node, or continue if you just need to make sure all nodes are added or whatever) - add a const ErrNodeAlreadyInSnapshot?

Reply:
I do not like the idea of comparing errors with constant. If we want to have a notion of error type we should explicitly model that.
And then return 'ClusterSnapshotError' which exposes `GetErrorType()` instead of `error`.

If you feel that it is beneficial we can do it as a followup, though I am not 100% convinced.

Reply:
More complex typed errors are really useful if they carry error type (that can be checked against) and some useful additional data. In this case there is no extra data to be returned - if I call AddNode(x) and get ErrNodeAlreadyInSnapshot I already have full information. Personally I find integers to be more intuitive for this kind of stuff, but I don't really see any practical difference between const int and const string in this case (and string seems to be more inline with golang coding style).

We already use compare-to-const error checking in cloudprovider (ErrNotImplemented, etc). In this case it's technically a const struct (which happen to be a typed error), but we really use it in the same way as we would a const string and const int.

Reply:
We discussed it offline with @losipiuk, I added changing this to the list of follow-ups to do in next PRs.